optaplanner-website-root/content/blog/2022/03/04/Real-time planning meets the SolverManager.adoc

```diff
@@ -0,0 +1,197 @@
+
+Every time OptaPlanner finds a new best solution, a component called `SolutionCloner` clones the working solution, which, in contrast to the best solution,
+keeps changing until the solving terminates. However, for performance reasons, OptaPlanner does not make a deep clone of the working solution;
+it clones only those parts of the solution that change during solving - the planning entities.
+
+Other instances, called problem facts, are not cloned and thus the working solution and all the best solutions found during solving share them.
+Normally, this is right and intentional. Not if the problem facts change as a part of your `ProblemChange` implementation, though.
+In that case, any problem fact or a problem fact collection must be cloned first, otherwise you may corrupt your previous best solutions.
+
+For more details about solution cloning, please read the https://www.optaplanner.org/docs/optaplanner/latest/planner-configuration/planner-configuration.html#cloningASolution[documentation].
+
+In the next example, we are removing a `CloudComputer`, which is a problem fact:
+
+[source,java]
+----
+public class DeleteComputerProblemChange implements ProblemChange<CloudBalance> {
+
+    private final CloudComputer computer;
+
+    public DeleteComputerProblemChange(CloudComputer computer) {
+        this.computer = computer;
+    }
+
+    @Override
+    public void doChange(CloudBalance cloudBalance, ProblemChangeDirector problemChangeDirector) {
+        CloudComputer workingComputer = problemChangeDirector.lookUpWorkingObjectOrFail(computer); // <1>
+        for (CloudProcess process : cloudBalance.getProcessList()) {
+            if (process.getComputer() == workingComputer) {
+                problemChangeDirector.changeVariable(process, "computer",
+                    workingProcess -> workingProcess.setComputer(null)); // <2>
+            }
+        }
+        List<CloudComputer> computerList = new ArrayList<>(cloudBalance.getComputerList()); // <3>
+        cloudBalance.setComputerList(computerList); // <3>
+        problemChangeDirector.removeProblemFact(workingComputer, computerList::remove); // <4>
+    }
+}
+----
+<1> Finds the working solution counterpart of the `computer`. The `CloudComputer` must have a field annotated with `@PlanningId`.
+<2> Unassigns this computer from every process that runs on it. The string "computer" is the name of a `@PlanningVariable` field of the `CloudProcess`.
+<3> As the `SolutionCloner` does not clone a problem fact collection, it has to be done manually.
+<4> Removes the computer from the `computerList`.
+
+Important to note, the lookup (1) is required only because we need to search for the processes assigned to this computer.
+Every method that either removes or changes an object that is supposed to already exist in the working solution
+does the lookup internally. Thus, if we passed the `computer` instance to the `removeProblemFact()` (4) it would still work correctly:
```

Comment:
Is it really important to note? To me, it is just an implementation detail and is mudying the explanation.
The API is what it is, we do not need to explain what it could be.

Reply:
I guess you are right; the lookup is clearly mentioned in Javadoc of every method that performs it. That should be enough.